bigtable/metrics.go

```diff
@@ -0,0 +1,487 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bigtable
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/bigtable/internal"
+	"github.com/google/uuid"
+	gax "github.com/googleapis/gax-go/v2"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"google.golang.org/api/option"
+	btpb "google.golang.org/genproto/googleapis/bigtable/v2"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	builtInMetricsMeterName = "bigtable.googleapis.com/internal/client/"
+
+	metricsPrefix         = "bigtable/"
+	locationMDKey         = "x-goog-ext-425905942-bin"
+	serverTimingMDKey     = "server-timing"
+	serverTimingValPrefix = "gfet4t7; dur="
+
+	// Monitored resource labels
+	monitoredResLabelKeyProject  = "project_id"
+	monitoredResLabelKeyInstance = "instance"
+	monitoredResLabelKeyTable    = "table"
+	monitoredResLabelKeyCluster  = "cluster"
+	monitoredResLabelKeyZone     = "zone"
+
+	// Metric labels
+	metricLabelKeyAppProfile         = "app_profile"
+	metricLabelKeyMethod             = "method"
+	metricLabelKeyOperationStatus    = "status"
+	metricLabelKeyStreamingOperation = "streaming"
+	metricLabelKeyClientName         = "client_name"
+	metricLabelKeyClientUID          = "client_uid"
+
+	// Metric names
+	metricNameOperationLatencies = "operation_latencies"
+	metricNameAttemptLatencies   = "attempt_latencies"
+	metricNameServerLatencies    = "server_latencies"
+	metricNameRetryCount         = "retry_count"
+)
+
+var (
+	// duration between two metric exports
+	defaultSamplePeriod = 5 * time.Minute
+
+	clientName            = fmt.Sprintf("cloud.google.com/go/bigtable v%v", internal.Version)
+	builtInEnabledDefault = true
+
+	bucketBounds = []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0, 10.0, 13.0, 16.0, 20.0, 25.0, 30.0, 40.0,
+		50.0, 65.0, 80.0, 100.0, 130.0, 160.0, 200.0, 250.0, 300.0, 400.0, 500.0, 650.0,
+		800.0, 1000.0, 2000.0, 5000.0, 10000.0, 20000.0, 50000.0, 100000.0, 200000.0,
+		400000.0, 800000.0, 1600000.0, 3200000.0}
+
+	// All the built-in metrics have same attributes except 'status' and 'streaming'
+	// These attributes need to be added to only few of the metrics
+	additionalAttributes = map[string][]string{
+		metricNameOperationLatencies: {
+			metricLabelKeyOperationStatus,
+			metricLabelKeyStreamingOperation,
+		},
+		metricNameAttemptLatencies: {
+			metricLabelKeyOperationStatus,
```

Comment:
this is a bit weird to see OperationStatus in attempt latencies. Maybe the const key should be renamed to be applicable to both? (same for serverLatencies below). Or better still, there should be a separate const for attemptStatus

Reply:
Updated